cpp/src/strings/utilities.cu

```diff
@@ -147,9 +148,30 @@ std::mutex g_flags_table_mutex;
 std::mutex g_cases_table_mutex;
 std::mutex g_special_case_mappings_mutex;
```

Comment:
since we have a class now, the mutex could reside inside the `per_context_cache` class.

Reply:
> since we have a class now, the mutex could reside inside the `per_context_cache` class.

I agree with this. The `std::unordered_map` is not thread-safe so I think you could reduce this down to a single mutex to protect the map and its contents.

Reply:
I intentionally made the `per_context_cache` thread unsafe so that the user can choose a strategy to make it thread safe. For example, if libcudf had a proper bootstrapping sequence that initializes this cache and other global states like this before any user touches, we don't even need to hold mutex to read this cache later (AFAIK, `std::unordered_map` is safe for concurrent read accesses). And that bootstrapping logic could take a mutex once for the whole initialization steps for each CUDA context.

If you still think that I should push down the mutex to the `per_context_cache`, let me know.

Reply:
I eventually decided to make and use a thread safe version for now. Please take a look at my latest commit.